### PR TITLE
[REFACTOR] OAuthManager + increase test coverage

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

Codecov is having some problems with this branch so here is the code coverage so far:  

Diff coverage: 95.18%
Total coverage: 17.86% (+ ~2%)

### CODE REVIEW

File: Rocket.Chat/Controllers/Auth/WebViewController.swift [hunk 1]
```diff
@@ -19,6 +19,12 @@ class WebViewController: OAuthWebViewController {
 
     var activityIndicator: UIActivityIndicatorView!
 
+    convenience init(authorizeURL: URL?) {
+        self.init()
+        self.targetURL = authorizeURL
+        self.viewDidLoad()
```
Comment (artrmz) [inline Rocket.Chat/Controllers/Auth/WebViewController.swift]: Should we be calling it in init?

Comment (cardoso) [reply Rocket.Chat/Controllers/Auth/WebViewController.swift]: That's a workaround to ensure the WKWebView is loaded, unfortunately :)

I'll see what I can do.

File: Rocket.Chat/OAuthManager.swift [hunk 2]
```diff
@@ -31,72 +31,69 @@ class OAuthCredentials {
...
 
+        let fragmentJSON = JSON(parseJSON: normalizedFragment)
+        return OAuthCredentials(json: fragmentJSON)
+    }
+
+    static func callbackUrl(for loginService: LoginService, server: URL) -> URL? {
+        guard
+            let host = server.host,
+            let service = loginService.service
+            else {
```
Comment (rafaelks) [inline Rocket.Chat/OAuthManager.swift]: 🔙

File: Rocket.Chat/OAuthManager.swift [hunk 3]
```diff
@@ -31,72 +31,69 @@ class OAuthCredentials {
...
-            let scope = loginService.scope,
-            let callbackURL = callbackURL(for: loginService, at: server),
-            let state = state()
-        else {
-            failure()
-            return
+            let authorizeUrl = loginService.authorizeUrl,
+            let accessTokenUrl = loginService.accessTokenUrl,
+            let clientId = loginService.clientId
+            else {
```
Comment (rafaelks) [inline Rocket.Chat/OAuthManager.swift]: 🔙

File: Rocket.ChatTests/Controllers/OAuthViewControllerSpec.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        let controller = OAuthViewController.testInstance()
+
+        XCTAssertTrue(controller.willNavigate(controller.webView, to: controller.callbackUrl, decisionHandler: { _ in }))
+        XCTAssertFalse(controller.willNavigate(controller.webView, to: nil, decisionHandler: { _ in }))
+
+        let url: URL! = URL(string: "https://a.com/#%7B%22credentialToken%22:%22token%22,%22credentialSecret%22:%22secret%22%7D")
+        controller.callbackUrl = url
+        XCTAssertTrue(controller.willNavigate(controller.webView, to: controller.callbackUrl, decisionHandler: { _ in }))
+    }
+}
```
Comment (rafaelks) [inline Rocket.ChatTests/Controllers/OAuthViewControllerSpec.swift]: ✅ ✅ ✅

File: Rocket.ChatTests/Managers/OAuthManagerSpec.swift [hunk 5]
```diff
@@ -7,26 +7,94 @@
...
+        }
+
+        XCTAssertEqual(credentials.secret, "dj_4JCx3Ix8fL4IEPnH-redacted", "token is correct")
+        XCTAssertEqual(credentials.token, "blvELQFjdP8q5u6Je0ceTYDChredacted", "secret is correct")
+    }
+
+    func testOAuthCredentialsFail() {
+        let json = JSON([
+            "credentialSecret": "dj_4JCx3Ix8fL4IEPnH-redacted"
+            ])
```
Comment (rafaelks) [inline Rocket.ChatTests/Managers/OAuthManagerSpec.swift]: 🔙

File: Rocket.ChatTests/Managers/OAuthManagerSpec.swift [hunk 6]
```diff
@@ -7,26 +7,94 @@
...
+        XCTAssertFalse(valid)
+    }
+
+    func testAuthorizeWithValidLoginService() {
+        let loginService = LoginService.testInstance()
+        let url: URL! = URL(string: "https://open.rocket.chat")
+        let valid = OAuthManager.authorize(loginService: loginService, at: url, viewController: UIViewController(), success: { _ in }, failure: { })
+
+        XCTAssertTrue(valid)
+    }
```
Comment (rafaelks) [inline Rocket.ChatTests/Managers/OAuthManagerSpec.swift]: ✅

File: Rocket.ChatTests/Models/LoginServiceSpec.swift [hunk 7]
```diff
@@ -71,4 +84,28 @@ class LoginServiceSpec: XCTestCase, RealmTestCase {
...
+        let service = LoginService()
+        service.serverUrl = "https://open.rocket.chat/"
+        service.tokenPath = "token_path"
+
+        XCTAssertEqual(service.accessTokenUrl, "https://open.rocket.chat/token_path")
+
+        service.tokenPath = nil
+
+        XCTAssertNil(service.accessTokenUrl, "https://open.rocket.chat/token_path")
+    }
```
Comment (rafaelks) [inline Rocket.ChatTests/Models/LoginServiceSpec.swift]: ✅

**Pull request merged**